File: configuration-manager-go-client/http_client.go
package configmanager

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"mime/multipart"
	"net/http"
	"net/url"
	"strings"
	"time"
)

// HTTPClient provides HTTP operations with authentication
type HTTPClient struct {
	baseURL    string
	apiKey     string
	jwtToken   string
	httpClient *http.Client
	userAgent  string
}

// NewHTTPClient creates a new HTTP client with the specified options
func NewHTTPClient(options *ClientOptions) *HTTPClient {
	httpClient := options.HTTPClient
	if httpClient == nil {
		httpClient = &http.Client{
			Timeout: options.Timeout,
		}
	}

	return &HTTPClient{
		baseURL:    strings.TrimSuffix(options.BaseURL, "/"),
		apiKey:     options.APIKey,
		jwtToken:   options.JWTToken,
		httpClient: httpClient,
		userAgent:  options.UserAgent,
	}
}

// SetJWTToken updates the JWT token for authentication
func (c *HTTPClient) SetJWTToken(token string) {
	c.jwtToken = token
}

// SetAPIKey updates the API key for authentication
func (c *HTTPClient) SetAPIKey(apiKey string) {
	c.apiKey = apiKey
}

// Get performs a GET request
func (c *HTTPClient) Get(ctx context.Context, path string) (*http.Response, error) {
	return c.do(ctx, http.MethodGet, path, nil, nil)
}

// GetWithQuery performs a GET request with query parameters
func (c *HTTPClient) GetWithQuery(ctx context.Context, path string, params url.Values) (*http.Response, error) {
	if len(params) > 0 {
		path = path + "?" + params.Encode()
	}
	return c.Get(ctx, path)
}

// Post performs a POST request with JSON body
func (c *HTTPClient) Post(ctx context.Context, path string, body interface{}) (*http.Response, error) {
	return c.do(ctx, http.MethodPost, path, body, map[string]string{
		"Content-Type": "application/json",
	})
}

// PostForm performs a POST request with form data
func (c *HTTPClient) PostForm(ctx context.Context, path string, data url.Values) (*http.Response, error) {
	return c.do(ctx, http.MethodPost, path, strings.NewReader(data.Encode()), map[string]string{
		"Content-Type": "application/x-www-form-urlencoded",
	})
}

// PostMultipart performs a POST request with multipart form data
func (c *HTTPClient) PostMultipart(ctx context.Context, path string, body *bytes.Buffer, contentType string) (*http.Response, error) {
	return c.do(ctx, http.MethodPost, path, body, map[string]string{
		"Content-Type": contentType,
	})
}

// Put performs a PUT request with JSON body
func (c *HTTPClient) Put(ctx context.Context, path string, body interface{}) (*http.Response, error) {
	return c.do(ctx, http.MethodPut, path, body, map[string]string{
		"Content-Type": "application/json",
	})
}

// Delete performs a DELETE request
func (c *HTTPClient) Delete(ctx context.Context, path string) (*http.Response, error) {
	return c.do(ctx, http.MethodDelete, path, nil, nil)
}

// do performs the actual HTTP request
func (c *HTTPClient) do(ctx context.Context, method, path string, body interface{}, headers map[string]string) (*http.Response, error) {
	// Build URL
	fullURL := c.baseURL + path

	// Prepare body
	var reqBody io.Reader
	if body != nil {
		switch v := body.(type) {
		case io.Reader:
			reqBody = v
		case string:
			reqBody = strings.NewReader(v)
		case []byte:
			reqBody = bytes.NewReader(v)
		default:
			// JSON encode
			jsonData, err := json.Marshal(body)
			if err != nil {
				return nil, fmt.Errorf("failed to marshal request body: %w", err)
			}
			reqBody = bytes.NewReader(jsonData)
		}
	}

	// Create request
	req, err := http.NewRequestWithContext(ctx, method, fullURL, reqBody)
	if err != nil {
		return nil, NewNetworkError(fmt.Sprintf("failed to create request: %v", err), err)
	}

	// Set headers
	c.setHeaders(req, headers)

	// Perform request
	resp, err := c.httpClient.Do(req)
	if err != nil {
		if ctx.Err() == context.DeadlineExceeded {
			return nil, NewTimeoutError("request timed out", c.httpClient.Timeout)
		}
		return nil, NewNetworkError(fmt.Sprintf("request failed: %v", err), err)
	}

	// Check for HTTP errors
	if resp.StatusCode >= 400 {
		return resp, c.handleErrorResponse(resp)
	}

	return resp, nil
}

// setHeaders sets the appropriate headers for the request
func (c *HTTPClient) setHeaders(req *http.Request, additionalHeaders map[string]string) {
	// Set User-Agent
	if c.userAgent != "" {
		req.Header.Set("User-Agent", c.userAgent)
	}

	// Set authentication headers
	if c.jwtToken != "" {
		req.Header.Set("Authorization", "Bearer "+c.jwtToken)
	} else if c.apiKey != "" {
		req.Header.Set("X-API-Key", c.apiKey)
	}

	// Set additional headers
	for key, value := range additionalHeaders {
		req.Header.Set(key, value)
	}

	// Set Accept header if not already set
	if req.Header.Get("Accept") == "" {
		req.Header.Set("Accept", "application/json")
	}
}

// handleErrorResponse handles HTTP error responses and creates appropriate error types
func (c *HTTPClient) handleErrorResponse(resp *http.Response) error {
	defer resp.Body.Close()

	// Read response body
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return NewHTTPError(fmt.Sprintf("failed to read error response: %v", err), nil, resp)
	}

	// Try to parse as API error response
	var apiError APIErrorResponse
	if err := json.Unmarshal(body, &apiError); err != nil {
		// If parsing fails, use raw body as error message
		apiError.Error = string(body)
	}

	message := apiError.Error
	if message == "" {
		message = fmt.Sprintf("HTTP %d error", resp.StatusCode)
	}

	// Create specific error types based on status code
	switch resp.StatusCode {
	case http.StatusUnauthorized:
		return NewAuthenticationError(message, resp.StatusCode)
	case http.StatusForbidden:
		return NewAuthorizationError(message, resp.StatusCode)
	case http.StatusNotFound:
		return NewNotFoundError(message, resp.StatusCode)
	case http.StatusBadRequest:
		return NewValidationError(message, resp.StatusCode)
	case http.StatusConflict:
		return NewConflictError(message, resp.StatusCode)
	case http.StatusTooManyRequests:
		// Default retry after 1 minute if not specified
		retryAfter := time.Now().Add(time.Minute)
		return NewRateLimitError(message, resp.StatusCode, &retryAfter)
	default:
		return NewAPIError(message, resp.StatusCode, &apiError)
	}
}

// DecodeJSON decodes JSON response body into the provided interface
func (c *HTTPClient) DecodeJSON(resp *http.Response, v interface{}) error {
	defer resp.Body.Close()

	if err := json.NewDecoder(resp.Body).Decode(v); err != nil {
		return fmt.Errorf("failed to decode JSON response: %w", err)
	}

	return nil
}

// CreateMultipartForm creates a multipart form for file uploads
func (c *HTTPClient) CreateMultipartForm() (*bytes.Buffer, *multipart.Writer) {
	body := &bytes.Buffer{}
	writer := multipart.NewWriter(body)
	return body, writer
}

// CloseResponse safely closes a response body
func (c *HTTPClient) CloseResponse(resp *http.Response) {
	if resp != nil && resp.Body != nil {
		resp.Body.Close()
	}
}

File: configuration-manager-python-client/configuration_manager_client/http_client.py
"""Base HTTP client for Configuration Manager."""

import json
from typing import Optional, Dict, Any, Union, BinaryIO
from urllib.parse import urljoin, urlencode

import requests
from requests.adapters import HTTPAdapter
from urllib3.util.retry import Retry

from .exceptions import (
    create_error_from_response,
    create_network_error,
    create_timeout_error,
)


class HTTPClient:
    """HTTP client with authentication for Configuration Manager API."""

    def __init__(
        self,
        base_url: str,
        api_key: Optional[str] = None,
        jwt_token: Optional[str] = None,
        timeout: float = 30.0,
        max_file_size: int = 50 * 1024 * 1024,  # 50MB
        user_agent: str = "configuration-manager-python-client/2.0.0",
        verify_ssl: bool = True,
        max_retries: int = 3,
    ):
        """Initialize HTTP client.
        
        Args:
            base_url: Base URL of the Configuration Manager API
            api_key: API key for authentication (optional)
            jwt_token: JWT token for authentication (optional)
            timeout: Request timeout in seconds
            max_file_size: Maximum file size for uploads
            user_agent: User agent string
            verify_ssl: Whether to verify SSL certificates
            max_retries: Maximum number of retries for failed requests
        """
        self.base_url = base_url.rstrip("/")
        self.api_key = api_key
        self.jwt_token = jwt_token
        self.timeout = timeout
        self.max_file_size = max_file_size
        self.user_agent = user_agent
        self.verify_ssl = verify_ssl

        # Create session with retry strategy
        self.session = requests.Session()
        
        # Configure retry strategy
        retry_strategy = Retry(
            total=max_retries,
            status_forcelist=[429, 500, 502, 503, 504],
            backoff_factor=1,
            raise_on_status=False,
        )
        adapter = HTTPAdapter(max_retries=retry_strategy)
        self.session.mount("http://", adapter)
        self.session.mount("https://", adapter)

        # Set default headers
        self._update_headers()

    def _update_headers(self) -> None:
        """Update session headers with authentication and common headers."""
        self.session.headers.clear()
        
        # Set User-Agent
        if self.user_agent:
            self.session.headers["User-Agent"] = self.user_agent

        # Set authentication headers
        if self.jwt_token:
            self.session.headers["Authorization"] = f"Bearer {self.jwt_token}"
        elif self.api_key:
            self.session.headers["X-API-Key"] = self.api_key

        # Set common headers
        self.session.headers["Accept"] = "application/json"

    def set_jwt_token(self, token: str) -> None:
        """Update JWT token for authentication."""
        self.jwt_token = token
        self._update_headers()

    def set_api_key(self, api_key: str) -> None:
        """Update API key for authentication."""
        self.api_key = api_key
        self._update_headers()

    def _build_url(self, path: str) -> str:
        """Build full URL from path."""
        return urljoin(self.base_url, path.lstrip("/"))

    def _handle_response(self, response: requests.Response) -> requests.Response:
        """Handle response and raise appropriate exceptions for errors."""
        if response.status_code >= 400:
            raise create_error_from_response(response)
        return response

    def get(
        self,
        path: str,
        params: Optional[Dict[str, Any]] = None,
        headers: Optional[Dict[str, str]] = None,
    ) -> requests.Response:
        """Perform GET request."""
        url = self._build_url(path)
        
        try:
            response = self.session.get(
                url,
                params=params,
                headers=headers,
                timeout=self.timeout,
                verify=self.verify_ssl,
            )
            return self._handle_response(response)
        except requests.exceptions.Timeout:
            raise create_timeout_error(f"Request to {url} timed out", self.timeout)
        except requests.exceptions.RequestException as e:
            raise create_network_error(f"Request to {url} failed: {e}", e)

    def post(
        self,
        path: str,
        data: Optional[Union[Dict[str, Any], str, bytes]] = None,
        json_data: Optional[Dict[str, Any]] = None,
        files: Optional[Dict[str, Any]] = None,
        headers: Optional[Dict[str, str]] = None,
    ) -> requests.Response:
        """Perform POST request."""
        url = self._build_url(path)

        # Set content type for JSON if not uploading files
        if json_data is not None and files is None:
            if headers is None:
                headers = {}
            headers.setdefault("Content-Type", "application/json")

        try:
            response = self.session.post(
                url,
                data=data,
                json=json_data,
                files=files,
                headers=headers,
                timeout=self.timeout,
                verify=self.verify_ssl,
            )
            return self._handle_response(response)
        except requests.exceptions.Timeout:
            raise create_timeout_error(f"Request to {url} timed out", self.timeout)
        except requests.exceptions.RequestException as e:
            raise create_network_error(f"Request to {url} failed: {e}", e)

    def put(
        self,
        path: str,
        data: Optional[Union[Dict[str, Any], str, bytes]] = None,
        json_data: Optional[Dict[str, Any]] = None,
        headers: Optional[Dict[str, str]] = None,
    ) -> requests.Response:
        """Perform PUT request."""
        url = self._build_url(path)

        # Set content type for JSON
        if json_data is not None:
            if headers is None:
                headers = {}
            headers.setdefault("Content-Type", "application/json")

        try:
            response = self.session.put(
                url,
                data=data,
                json=json_data,
                headers=headers,
                timeout=self.timeout,
                verify=self.verify_ssl,
            )
            return self._handle_response(response)
        except requests.exceptions.Timeout:
            raise create_timeout_error(f"Request to {url} timed out", self.timeout)
        except requests.exceptions.RequestException as e:
            raise create_network_error(f"Request to {url} failed: {e}", e)

    def delete(
        self,
        path: str,
        headers: Optional[Dict[str, str]] = None,
    ) -> requests.Response:
        """Perform DELETE request."""
        url = self._build_url(path)

        try:
            response = self.session.delete(
                url,
                headers=headers,
                timeout=self.timeout,
                verify=self.verify_ssl,
            )
            return self._handle_response(response)
        except requests.exceptions.Timeout:
            raise create_timeout_error(f"Request to {url} timed out", self.timeout)
        except requests.exceptions.RequestException as e:
            raise create_network_error(f"Request to {url} failed: {e}", e)

    def download(
        self,
        path: str,
        headers: Optional[Dict[str, str]] = None,
    ) -> requests.Response:
        """Download file content."""
        url = self._build_url(path)

        try:
            response = self.session.get(
                url,
                headers=headers,
                timeout=self.timeout,
                verify=self.verify_ssl,
                stream=True,  # Stream for large files
            )
            return self._handle_response(response)
        except requests.exceptions.Timeout:
            raise create_timeout_error(f"Download from {url} timed out", self.timeout)
        except requests.exceptions.RequestException as e:
            raise create_network_error(f"Download from {url} failed: {e}", e)

    def close(self) -> None:
        """Close the HTTP session."""
        if self.session:
            self.session.close()

    def __enter__(self):
        """Context manager entry."""
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        """Context manager exit."""
        self.close()
